posts/2021/05/may-status-update.md

```diff
@@ -0,0 +1,101 @@
+<!--
+.. title: HPy Status Update
+.. slug: may-status-update
+.. date: 2021-05-07 10:00:00 UTC
+.. author: hodgestar
+.. tags:
+.. category:
+.. link:
+.. description:
+.. type: text
+-->
+
+
+HPy's first git commit was on Friday, 12 July 2019. Despite the crazy state
+of the world, HPy has come a long way since then. This is a quick update
+to say where we are now and what's planned for the immediate future.
+
+<!--TEASER_END-->
+
+What can one do with HPy right now?
+===================================
+
+One can already port non-trivial Python extensions to the HPy API, but
+there is also still a lot of work to do. Things one can do right now
+with HPy include:
+
+  - Initialize an extension module
+  - Create custom extension types
+  - Create custom methods and functions
+  - Interact with many built-in types (including dictionaries, tuples,
+    lists, strings, bytes and longs)
+  - Raise and check exceptions
+  - Parse arguments to C functions
+  - Call Python functions
+  - Compile an extension in either "universal" or "cpython" modes (see below)
+  - Put a "universal" extension into debug mode **at runtime** (see below
+    for how this is possible without a performance penalty).
+  - Compile extensions for POSIX and Windows (MSVC).
+
+An extension written using HPy may be compiled either in "cpython" or
+"universal" mode:
+
+  - "cpython" mode produces an ordinary Python C extension that only uses
+    the HPy API internally. From the point of view of the Python implementation
+    it is an ordinary Python C extension.
+
+  - "universal" mode produces an HPy C extension. This extension does not use
+    the Python C API directly at all. A Python implementation must know how
+    to load such an extension and pass it an HPy context. Currently context
+    implementations exist for CPython, PyPy and GraalPython.
+
+Debug mode is implemented by replacing the HPy context with a debug mode
+context at runtime. The debug context wraps the original context and tracks
+all HPy API calls. Debug mode already supports finding handles that are
+leaked (i.e. not closed) and handles that are used after being closed.
+Additional debug features should now be fairly straightforward to add.
+
+The debug mode is written using HPy itself, so its implementation can be
+shared by different Python implementations that support HPy.
+
```

Comment:
We could mention `ujson_hpy` and `piconumpy_hpy` here as examples.

Reply:
Maybe @fangerer's suggestion somehow ended up on the wrong lines?

Reply:
I've left links to ultrajson and piconumpy out because I'm not sure enough of the state of them relative to the most recent HPy to recommend that people look at them without talking to us first.